pupil_src/shared_modules/surface_tracker/surface.py

```diff
@@ -0,0 +1,513 @@
+        if len(points.shape) == 1:
+            points[1] = 1 - points[1]
+        else:
+            points[:, 1] = 1 - points[:, 1]
+        points *= camera_model.resolution
+        return points
+
+    def _finalize_def(self):
+        """
+        - prune markers that have been visible in less than x percent of frames.
+        - of those markers select a good subset of uv coords and compute mean.
+        - this mean value will be used from now on to estable surface transform
+        """
+        persistent_markers = {}
+        persistent_markers_dist = {}
+        for (k, m), m_dist in zip(
+            self.reg_markers_undist.items(), self.reg_markers_dist.values()
+        ):
+            if len(m.observations) > self._required_obs_per_marker * .5:
+                persistent_markers[k] = m
+                persistent_markers_dist[k] = m_dist
+        self.reg_markers_undist = persistent_markers
+        self.reg_markers_dist = persistent_markers_dist
+
+    def move_corner(self, corner_idx, new_pos, camera_model):
+        """Update surface definition by moving one of the corners to a new position.
+
+        Args:
+            corner_idx: Identifier of the corner to be moved. The order of corners is
+            `[(0, 0), (1, 0), (1, 1), (0, 1)]`
+            new_pos: The updated position of the corner in image pixel coordinates.
+            camera_model: Camera Model object.
+
+        """
+        # Markers undistorted
+        new_corner_pos = self.map_to_surf(
+            new_pos, camera_model, compensate_distortion=True
+        )
+        old_corners = np.array([(0, 0), (1, 0), (1, 1), (0, 1)], dtype=np.float32)
+
+        new_corners = old_corners.copy()
+        new_corners[corner_idx] = new_corner_pos
+
+        transform = cv2.getPerspectiveTransform(new_corners, old_corners)
+        for m in self.reg_markers_undist.values():
+            m.verts_uv = cv2.perspectiveTransform(
+                m.verts_uv.reshape((-1, 1, 2)), transform
+            ).reshape((-1, 2))
+
+        # Markers distorted
+        new_corner_pos = self.map_to_surf(
+            new_pos, camera_model, compensate_distortion=False
+        )
+        old_corners = np.array([(0, 0), (1, 0), (1, 1), (0, 1)], dtype=np.float32)
+
+        new_corners = old_corners.copy()
+        new_corners[corner_idx] = new_corner_pos
+
+        transform = cv2.getPerspectiveTransform(new_corners, old_corners)
+        for m in self.reg_markers_dist.values():
+            m.verts_uv = cv2.perspectiveTransform(
+                m.verts_uv.reshape((-1, 1, 2)), transform
+            ).reshape((-1, 2))
+
+    def add_marker(self, marker_id, verts_px, camera_model):
+        """Add a marker to the surface definition."""
+        surface_marker_dist = _Surface_Marker_Aggregate(marker_id)
+        marker_verts_dist = np.array(verts_px).reshape((4, 2))
+        uv_coords_dist = self.map_to_surf(
+            marker_verts_dist, camera_model, compensate_distortion=False
```

Comment:
Are you sure about `compensate_distortion=False`? Looks like code duplication similar to above

Reply:
That is a bug! Good find!